### PR TITLE
fix: A bug fix for QueryDeepPartialEntity type definition to allow pr…

### LINKED ISSUE #10155: Problem with QueryDeepPartialEntity type after commit a1960e11
### Issue description

It's impossible to build abstraction on top of repositories with generics 

### Expected Behavior

I want to build a custom abstract repo that will override each method. But I can't build it because of the definition of this type QueryDeepPartialEntity

```typescript
import { EntityHelper } from '@libs/core-modules/database/common/entity-helper';
import {
  Column,
  CreateDateColumn,
  DataSource,
  EntityTarget,
  InsertResult,
  Repository,
  UpdateDateColumn,
  VersionColumn,
} from 'typeorm';
import { QueryDeepPartialEntity } from 'typeorm/query-builder/QueryPartialEntity';

abstract class BaseEntityHelper extends EntityHelper {
  id!: string | number;

  @CreateDateColumn()
  createdAt!: Date;

  @UpdateDateColumn()
  updatedAt!: Date;

  @Column({ nullable: true, type: 'datetime' })
  deletedAt?: Date;

  @VersionColumn()
  version!: number;
}

export default abstract class BaseRepository<
  ENTITY extends BaseEntityHelper,
> extends Repository<ENTITY> {
  protected constructor(
    entityTarget: EntityTarget<ENTITY>,
    private dataSource: DataSource,
  ) {
    super(entityTarget, dataSource.createEntityManager());
  }

  insert(
    entity: QueryDeepPartialEntity<ENTITY> | QueryDeepPartialEntity<ENTITY>[],
  ): Promise<InsertResult> {
    // error here, property doesn't exist
    // but it should be here because ENTITY must extend BaseEntityHelper
    entity.createdAt;

    return super.insert(entity);
  }
}

```

### Actual Behavior

Access obviously defined fields when using the inheritance model 

### Steps to reproduce

I attached a code with a problem in the expected behavior section 

### My Environment

| Dependency          | Version  |
| ---                 | ---      |
| Operating System    |          |
| Node.js version     | v18.16.0  |
| Typescript version  | 4.9.5  |
| TypeORM version     | 0.3.16  |

### Additional Context

I found that the problem with this type 

```typescript
export type QueryDeepPartialEntity<T> = _QueryDeepPartialEntity<
    ObjectLiteral extends T ? unknown : T
>
```

When we pass a generic here, it's erasing the information about any types that it forces to extend, so it's deciding that my generic is an object literal, but it actually not.  

I think it makes sense to change it to this 

```typescript 

/**
 * Make all properties in T optional. Deep version.
 */
export type QueryDeepPartialEntity<T> = _QueryDeepPartialEntity<
    T extends "toString" ? T : unknown
>
```

I found a few other places where people distinguish between classes and objects so probably it makes sense to change it. 
But I need input from somebody who is more aware of the project. 

### Relevant Database Driver(s)

- [ ] aurora-mysql
- [ ] aurora-postgres
- [ ] better-sqlite3
- [ ] cockroachdb
- [ ] cordova
- [ ] expo
- [ ] mongodb
- [ ] mysql
- [ ] nativescript
- [ ] oracle
- [ ] postgres
- [ ] react-native
- [ ] sap
- [ ] spanner
- [ ] sqlite
- [ ] sqlite-abstract
- [ ] sqljs
- [ ] sqlserver

### Are you willing to resolve this issue by submitting a Pull Request?

Yes, I have the time, and I know how to start.

### PR DESCRIPTION
**Pull request by ghost**
### Description of change

Fixes #10155

I verified it by pulling this branch into my project and experimenting with a change. Now inheritance works properly 

### Pull-Request Checklist

- [ x ] Code is up-to-date with the `master` branch
- [ x ] `npm run format` to apply prettier formatting
- [ N/A ] `npm run test` passes with this change (can't run on M1)
- [ ] This pull request links relevant issues as `Fixes #10155`
- [ N/A ] There are new or updated unit tests validating the change (I don't think any new tests are needed, it just types) 
- [ N/A ] Documentation has been updated to reflect this change
- [ x ] The new commits follow conventions explained in [CONTRIBUTING.md](https://github.com/typeorm/typeorm/blob/master/CONTRIBUTING.md)

### CODE REVIEW

File: src/query-builder/QueryPartialEntity.ts [hunk 1]
```diff
@@ -11,7 +9,7 @@ export type QueryPartialEntity<T> = {
  * Make all properties in T optional. Deep version.
  */
 export type QueryDeepPartialEntity<T> = _QueryDeepPartialEntity<
-    ObjectLiteral extends T ? unknown : T
+    T extends "toString" ? T : unknown
```
Comment (hinogi) [inline src/query-builder/QueryPartialEntity.ts:12]: For me it seems like

```ts
/**
 * Make all properties in T optional
 */
export type QueryPartialEntity<T> = {
    [P in keyof T]?: T[P] | (() => string)
}
```

could be an issue.
I would have expect something like

```ts
/**
 * Make all properties in T optional
 */
export type QueryPartialEntity<T> = {
    [P in keyof T]?: QueryPartialEntity<T[K]>
}
```

for Deep partial objects, or the long version

```ts
interface DeepPartialArray<T> extends Array<DeepPartial<T>> {}

export type QueryPartialEntityObject<T> = {
    [P in keyof T]?: QueryPartialEntity<T[K]>
}
export type QueryPartialEntity<T> = T extends Function
  ? T
  : T extends Array<infer InferredArrayMember>
  ? DeepPartialArray<InferredArrayMember>
  : T extends ObjectLiteral
  ? QueryPartialEntityObject<T>
  : T | undefined;
```

props to @mattpocockuk

https://www.youtube.com/watch?v=AhzjPAtzGTs

**Review by hinogi — CHANGES_REQUESTED** [review]
I think the QueryPartialEntity needs some work first, or the comment needs more precision on what kind of partial object is expected and what members it could have

**Comment by kusigit** [comment]
`[ N/A ] npm run test passes with this change (can't run on M1)`

You have to enable `Use Rosetta for x86/amd64 emulation on Apple Silicon` in the Docker settings `Features in development`.

**Comment by pleerock** [comment]
I didn't get the purpose of these changes based on the issue description. Let's see if others report back the same issue. Closing for now.

**Comment by cameron-hicks** [comment]
@pleerock would you be open to revisiting this bugfix? I'm having a bug similar to https://github.com/typeorm/typeorm/issues/9826 and @hinogi's suggestion seems to fix it.

**Pull request closed without merging**